src/Servers/IIS/IIS/src/Core/IO/AsyncIOEngine.cs

```diff
@@ -35,13 +37,96 @@ public ValueTask<int> ReadAsync(Memory<byte> memory)
     }
 
     public ValueTask<int> WriteAsync(ReadOnlySequence<byte> data)
+    {
+        if (SegmentsOverChunksLimit(data))
+        {
+            return WriteDataOverChunksLimit(data);
+        }
+
+        return WriteDataAsync(data);
+    }
+
+    private ValueTask<int> WriteDataAsync(in ReadOnlySequence<byte> data)
     {
         var write = GetWriteOperation();
         write.Initialize(_handler, data);
         Run(write);
         return new ValueTask<int>(write, 0);
     }
 
+    /*
+    * In case the number of chunks is bigger than responseMaxChunks we need to make multiple calls 
+    * to the native api https://docs.microsoft.com/en-us/iis/web-development-reference/native-code-api-reference/ihttpresponse-writeentitychunks-method  
+    * Despite the documentation states that feeding the function with more than 65535 chunks will cause the function to throw an exception, 
+    * it actually seems that 65534 is the maximum number of chunks allowed.
+    * Also, there seems to be a problem when slicing a ReadOnlySequence on segment borders tracked here https://github.com/dotnet/runtime/issues/67607
+    * That's why we only allow 65533 chunks.  
+    */
+    private async ValueTask<int> WriteDataOverChunksLimit(ReadOnlySequence<byte> data)
+    {
+        var hr = 0;
+        ushort segmentsCount = 0;
+        var start = 0;
+        var length = 0;
+        int result;
+
+        foreach (var segment in data)
+        {
+            segmentsCount++;
+            length += segment.Length;
+
+            if (segmentsCount == responseMaxChunks)
+            {
+                result = await WriteDataAsync(data.Slice(start, length));
+
+                SaveResultIfError(result, ref hr);
+
+                segmentsCount = 0;
+                start = length;
```

Comment:
Nit: Instead of keeping track of start independently and rewalking the beginning ROS every time, it'd be a little more efficient to slice here:

```suggestion
                data = data.Slice(length);
```

It would be `data.Slice(start + length)` if start still existed, but start is always effectively 0 if we do this.

It'd be even more efficient to [ROS.TryGet directly like the enumerator's MoveNext() function does](https://github.com/dotnet/runtime/blob/74bd4850158022ae2ac10578c6b1df0400e2f7ac/src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.cs#L638-L646) to mutate the existing struct as you loop, but you'd still need to keep a copy of the original ROS until it's written so you can slice from the start postion.

I don't think you really need to do the `TryGet` thing, but I think the `data = data.Slice(length);` change is simple enough we should do it.

Reply:
That's neat, will do it, didn't think about it, thank you.